components/de_amp/browser/de_amp_service.cc

```diff
@@ -0,0 +1,102 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/de_amp/browser/de_amp_service.h"
+
+#include <string>
+
+#include "base/feature_list.h"
+#include "base/no_destructor.h"
+#include "brave/components/de_amp/browser/de_amp_pref_names.h"
+#include "brave/components/de_amp/common/features.h"
+#include "components/prefs/pref_registry_simple.h"
+#include "components/prefs/pref_service.h"
+#include "third_party/re2/src/re2/re2.h"
+
+namespace de_amp {
+
+// Check for "amp" or "⚡" in <html> tag
+// https://amp.dev/documentation/guides-and-tutorials/learn/spec/amphtml/?format=websites#ampd
+static const char kGetHtmlTagPattern[] = "(<\\s*html\\s.*>)";
+static const char kDetectAmpPattern[] = "(?:<.*\\s.*(amp|⚡)(?:\\s.*>|>|/>))";
+// Look for canonical link tag and get href
+// https://amp.dev/documentation/guides-and-tutorials/learn/spec/amphtml/?format=websites#canon
+static const char kFindCanonicalLinkTagPattern[] =
+    "(<\\s*link\\s.*rel=(?:\"|')canonical(?:\"|')(?:\\s.*>|>|/>))";
+static const char kFindCanonicalHrefInTagPattern[] =
+    "href=(?:\"|')(.*?)(?:\"|')";
+
+DeAmpService::DeAmpService(PrefService* prefs) : prefs_(prefs) {}
+
+DeAmpService::~DeAmpService() {}
+
+// static
+void DeAmpService::RegisterProfilePrefs(PrefRegistrySimple* registry) {
+  registry->RegisterBooleanPref(kDeAmpPrefEnabled, false);
+}
+
+void DeAmpService::ToggleDeAmp() {
+  const bool enabled = prefs_->GetBoolean(kDeAmpPrefEnabled);
+  prefs_->SetBoolean(kDeAmpPrefEnabled, !enabled);
+}
+
+void DeAmpService::DisableDeAmpForTest() {
+  prefs_->SetBoolean(kDeAmpPrefEnabled, false);
+}
+
+bool DeAmpService::IsEnabled() {
+  if (!base::FeatureList::IsEnabled(de_amp::features::kBraveDeAMP)) {
+    return false;
+  }
+
+  const bool enabled = prefs_->GetBoolean(kDeAmpPrefEnabled);
+  return enabled;
+}
+
+bool DeAmpService::VerifyCanonicalLink(const GURL canonical_link,
+                                      const GURL original_url) {
+  // Canonical URL should be a valid URL,
+  // be HTTP(S) and not be the same as original URL
+  return canonical_link.is_valid() && canonical_link.SchemeIsHTTPOrHTTPS() &&
+         canonical_link != original_url;
+}
+
+// If AMP page, find canonical link
+// canonical link param is populated if found
```

Comment:
please add `// static` comment (according to the CC)

Reply:
Can't we check for this with `npm run lint`?